File: C++/Dynamic Programming/editDistance.cpp
Edit Distance - GFG

class Solution {
  public:
    int editDistance(string s, string t) {
        // Code here
        int m = s.length();
        int n = t.length();
        
        int dp[m+1][n+1];
        
        for(int i = 0; i < m+1; i++){
            for(int j = 0; j < n+1; j++){
                
                if(i==0)                    // If first string empty, insert all element in 1st string or remove from 2nd string
                    dp[i][j] = j;
                
                else if(j==0)               // Same reason
                    dp[i][j] = i;
                
                else if(s[i-1] == t[j-1])   // IF equal no operation needed move to the next
                    dp[i][j] = dp[i-1][j-1];
                
                else                        // Need one operation and pick minimum from it
                    dp[i][j] = 1 + min(dp[i-1][j-1], min(dp[i][j-1], dp[i-1][j]));
            }
        }
        
        return dp[m][n];
    }
};

File: C++/Dynamic Programming/Revision 2025/EditDistance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)

        dp = [[0 for j in range(m+1)] for i in range(n+1)]

        # word2 subproblem is empty so either delete or insert. 
        # +1 operation in every cell
        for i in range(n+1):
            dp[i][0] = i
        
        # Same reason
        for j in range(m+1):
            dp[0][j] = j
        
        for i in range(1, n+1):
            for j in range(1, m+1):
                # No operation
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                
                else:
                    dp[i][j] = 1 + min(
                        dp[i-1][j-1],       # No Operation
                        dp[i][j-1],         # Insert
                        dp[i-1][j]          # Delete
                    )

        return dp[n][m]
